### PR TITLE
Implementation of ERC20Gateway.proveCogateway

### LINKED ISSUE #276: Implement ERC20Gateway::proveGateway
- This is an external function.
- This calls `MessageInbox.proveStorageAccount`

### PR DESCRIPTION
**Pull request by abhayks1**
PR addresses below points:
- Implementation of ERC20Gateway::proveGateway
- Unit tests with proof data

**Note:**
Method has been updated to ERC20Gateway.**proveGateway** which is aligned with ConsensusCogateway.proveConsensusGateway method naming convention. https://github.com/mosaicdao/mosaic-1/blob/develop/contracts/consensus-gateway/ConsensusCogateway.sol#L189
It makes sense since here we are proving ERC20Cogateway contract. 
I have updated the same in the ticket comment. 

Fixes #276

### CODE REVIEW

**Comment by deepesh-kn** [comment]
> PR addresses below points:
> 
> * Implementation of ERC20Gateway::proveCogateway
> * Unit tests with proof data
> 
> **Note:**
> Method has been updated to ERC20Gateway.**proveCogateway** which is aligned with ConsensusCogateway.proveConsensusGateway method naming convention. https://github.com/mosaicdao/mosaic-1/blob/develop/contracts/consensus-gateway/ConsensusCogateway.sol#L189
> It makes sense since here we are proving ERC20Cogateway contract.
> I have updated the same in the ticket comment.
> 
> Fixes #276

We can change the `ConsensusCogateway::proveConsensusGateway` to `ConsensusCogateway::proveGateway` instead of the proposed change, else you need to make separate services in the facilitator.

Also as we were discussing about moving the message bus and proveGateway function in to ERC20GatewayBase, at that time we can make the change related to ConsensusCogateway.

**Comment by abhayks1** [comment]
```
Also as we were discussing about moving the message bus and proveGateway function in to ERC20GatewayBase, at that time we can make the change related to ConsensusCogateway.
```
This will reduce lot of code duplication as proveGateway can be shared by ConsensusGateway/ConsensusCogateway/ERC20Gateway/ERC20Cogateway contracts 👍.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 1]
```diff
@@ -80,4 +80,32 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+     * @notice This method verifies merkle proof of ERC20Cogateway contract.
+     *
+     * @param _blockNumber Block number at which ERC20Cogateway contract
+     *                     address is to be proven.
+     * @param _rlpAccount RLP encoded account node object.
+     * @param _rlpParentNodes RLP encoded value of account proof node array.
+     *
+     * \post Emits GatewayProven event on successful execution.
+     * \post It calls MessageInbox.proveStorageAccount() with parameters
+     *       blockNumber, rlpAccountNode, rlpParentNodes.
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Gateway.sol]:
```
     * \post Calls `MessageInbox::proveStorageAccount()` function with
     *       `_blockNumber`, `_rlpAccountNode`, `_rlpParentNodes` as input
     *       parameters.
     * \post Emits `GatewayProven` event with parameters `address(this)` and
     *       `_blockNumber`.
     

```

File: contracts/erc20-gateway/ERC20GatewayBase.sol [hunk 2]
```diff
@@ -23,6 +23,12 @@ contract ERC20GatewayBase {
     using SafeMath for uint256;
 
 
+    /* Events */
+
+    /** Emitted when Gateway/Cogateway contract is proven. */
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20GatewayBase.sol]:
```suggestion
    /** Emitted when the remote gateway contract account is proven. */
```

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
Looks good, just some minor documentation changes.

File: contracts/erc20-gateway/ERC20Gateway.sol [hunk 3]
```diff
@@ -80,4 +80,34 @@ contract ERC20Gateway is MasterCopyNonUpgradable, MessageBus, ERC20GatewayBase {
...
+    )
+        external
+    {
+        MessageInbox.proveStorageAccount(
+            _blockNumber,
+            _rlpAccount,
+            _rlpParentNodes
+        );
+
+        emit GatewayProven(address(this), _blockNumber);
```
Comment (0xsarvesh) [inline contracts/erc20-gateway/ERC20Gateway.sol]: This should emit the `Cogateway` address.

File: contracts/erc20-gateway/ERC20GatewayBase.sol [hunk 4]
```diff
@@ -23,6 +23,12 @@ contract ERC20GatewayBase {
     using SafeMath for uint256;
 
 
+    /* Events */
+
+    /** Emitted when the remote gateway contract address is proven. */
+    event GatewayProven(address gateway, uint256 blockNumber);
```
Comment (0xsarvesh) [inline contracts/erc20-gateway/ERC20GatewayBase.sol]:
```suggestion
    event GatewayProven(address remoteGateway, uint256 blockNumber);
```

File: test/erc20_gateway/erc20_gateway/prove_gateway.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+      assert.strictEqual(
+        eventObject.args.gateway,
+        erc20Gateway.address,
+        'Gateway address is incorrect in the event.',
+      );
+
+      assert.isOk(
+        eventObject.args.blockNumber.eqn(ProveERC20CogatewayProof.blockNumber),
+        'Block number is incorrect in the event.',
```
Comment (0xsarvesh) [inline test/erc20_gateway/erc20_gateway/prove_gateway.js]: Please add the expected and actual block number in the reason string.

**Review by 0xsarvesh — DISMISSED** [review]
Some minor changes are required. 
Thanks 👍

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 1 above)
Comment (abhayks1) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done!

File: contracts/erc20-gateway/ERC20GatewayBase.sol (hunk 2 above)
Comment (abhayks1) [reply contracts/erc20-gateway/ERC20GatewayBase.sol]: Done!

File: contracts/erc20-gateway/ERC20Gateway.sol (hunk 3 above)
Comment (abhayks1) [reply contracts/erc20-gateway/ERC20Gateway.sol]: Done!

File: contracts/erc20-gateway/ERC20GatewayBase.sol (hunk 4 above)
Comment (abhayks1) [reply contracts/erc20-gateway/ERC20GatewayBase.sol]: Done!

File: test/erc20_gateway/erc20_gateway/prove_gateway.js (hunk 5 above)
Comment (abhayks1) [reply test/erc20_gateway/erc20_gateway/prove_gateway.js]: Done!

**Comment by abhayks1** [comment]
Available for review again 🍏

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**